readers/hdfs/src/main/java/com/criteo/hadoop/garmadon/hdfs/offset/HdfsOffsetComputer.java

```diff
@@ -8,82 +8,78 @@
+        } else {
+            this.offsetFilePatternGenerator = Pattern.compile(String.format("^(\\d+)\\.cluster=%s\\.(\\d+)$", kafkaCluster));
+        }
     }
 
     @Override
     public Map<Integer, Long> computeOffsets(Collection<Integer> partitionIds) throws IOException {
+        Set<Integer> partitionIdsSet = new HashSet<>(partitionIds);
         final RemoteIterator<LocatedFileStatus> filesIterator = fs.listFiles(basePath, true);
-        final Collection<String> fileNames = new ArrayList<>();
-        final Map<Integer, Long> result = new HashMap<>();
+        final Map<Integer, Long> result = partitionIds.stream().collect(Collectors.toMap(Function.identity(), i -> NO_OFFSET));
 
         while (filesIterator.hasNext()) {
-            fileNames.add(filesIterator.next().getPath().getName());
-        }
-
-        for (int partitionId: partitionIds) {
-            result.put(partitionId, computeOffset(partitionId, fileNames));
-        }
-
-        return result;
-    }
-
-    private long computeOffset(int partitionId, Collection<String> fileNames) {
-        final Pattern partitionBasedPattern = offsetFilePatternGenerator.apply(partitionId);
-        long highestOffset = NO_OFFSET;
-
-        for (String fileName: fileNames) {
-            final Matcher matcher = partitionBasedPattern.matcher(fileName);
-            if (!matcher.matches() || matcher.groupCount() < 1) continue;
-
-            final String offsetString = matcher.group(1);
-            final long offset;
-
-            try {
-                 offset = Long.valueOf(offsetString);
-            } catch (NumberFormatException e) {
-                LOGGER.info("Couldn't deviate a valid offset from '{}'", offsetString);
-                continue;
-            }
-
-            if (offset > highestOffset) {
-                highestOffset = offset;
+            String fileName = filesIterator.next().getPath().getName();
+            Matcher matcher = offsetFilePatternGenerator.matcher(fileName);
+            if (matcher.matches()) {
+                try {
+                    int partitionId = Integer.parseInt(matcher.group(1));
+                    Long offset = Long.parseLong(matcher.group(2));
+                    if (partitionIdsSet.contains(partitionId)) {
+                        result.merge(partitionId, offset, Long::max);
+                    }
+                } catch (NumberFormatException e) {
+                    LOGGER.info("Couldn't deviate a valid offset from '{}'", fileName);
+                }
             }
         }
-
-        return highestOffset;
+        return result;
     }
 
     @Override
     public String computePath(LocalDateTime time, Offset offset) {
-        return String.format("%s/%d.%d", time.format(DateTimeFormatter.ISO_DATE), offset.getPartition(),
-                offset.getOffset());
+        if (kafkaCluster == null) {
+            return String.format("%s/%d.%d", time.format(DateTimeFormatter.ISO_DATE), offset.getPartition(),
```

Comment:
can we rely on injection at construction and avoid null checking ?

Reply:
done.